2019-May-Season/SoftUni-Information-Services/src/SIS.HTTP/Headers/HttpHeaderCollection.cs

```diff
@@ -1,39 +1,61 @@
-﻿using System.Collections.Generic;
-using System.Linq;
-using SIS.HTTP.Common;
-using SIS.HTTP.Headers.Contracts;
-
-namespace SIS.HTTP.Headers
+﻿namespace SIS.HTTP.Headers
 {
-    public class HttpHeaderCollection : IHttpHeaderCollection
+    using System.Linq;
+    using System.Collections;
+    using System.Collections.Generic;
+
+    using SIS.HTTP.Common;
+    using SIS.HTTP.Headers.Contracts;
+    public class HttpHeaderCollection : IHttpHeaderCollection, IEnumerable<IHttpHeader>
     {
-        private Dictionary<string, HttpHeader> httpHeaders;
+        private readonly IList<IHttpHeader> httpHeaders;
 
         public HttpHeaderCollection()
         {
-            this.httpHeaders = new Dictionary<string, HttpHeader>();
+            this.httpHeaders = new List<IHttpHeader>();
         }
 
-        public void AddHeader(HttpHeader header)
+        public void AddHeader(IHttpHeader header)
         {
             CoreValidator.ThrowIfNull(header, nameof(header));
-            this.httpHeaders.Add(header.Key, header);
+            this.httpHeaders.Add(header);
+        }
+
+        public string this[string key]
+        {
+            get
+            {
+                return this.httpHeaders.FirstOrDefault(h => h.Key == key).Value;
+            }
+            set
+            {
+                this[key] = value;
+            }
         }
 
         public bool ContainsHeader(string key)
         {
             CoreValidator.ThrowIfNullOrEmpty(key, nameof(key));
-            return this.httpHeaders.ContainsKey(key);
+            return this.httpHeaders.Any(h => h.Key == key);
```

Comment:
Using .Any() LINQ extension method here makes zero sense as it iterates over IEnumerable<T> and it's not scalable (O(n) iterations) while .ContainsKey() uses the hash code of the key, meaning that the complexity of the operation is close/equal to O(1). Using any defeats the purpose of fast searching using IDictionary<TKey, TVal> fast search by using it's hashtable.